src/storage/readpool_context.rs

```diff
@@ -0,0 +1,146 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt;
+use std::mem;
+use prometheus::local::{LocalCounterVec, LocalHistogramTimer, LocalHistogramVec};
+
+use server::readpool;
+use util::futurepool;
+use util::worker;
+use util::collections::HashMap;
+use pd;
+use storage;
+
+use super::metrics::*;
+
+pub struct Context {
+    pd_sender: Option<worker::FutureScheduler<pd::PdTask>>,
+
+    // TODO: command_duration, processing_read_duration can be merged together.
+    command_duration: LocalHistogramVec,
+    processing_read_duration: LocalHistogramVec,
+    command_keyreads: LocalHistogramVec,
+    // TODO: kv_command_counter, raw_command_counter, command_pri_counter can be merged together.
+    kv_command_counter: LocalCounterVec,
+    raw_command_counter: LocalCounterVec,
+    command_pri_counter: LocalCounterVec,
+    scan_details: LocalCounterVec,
+
+    read_flow_stats: HashMap<u64, storage::FlowStatistics>,
+}
+
+impl fmt::Debug for Context {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("storage::ReadPoolContext").finish()
+    }
+}
+
+impl Context {
+    pub fn new(pd_sender: Option<worker::FutureScheduler<pd::PdTask>>) -> Self {
+        Context {
+            pd_sender,
+            command_duration: SCHED_HISTOGRAM_VEC.local(),
+            processing_read_duration: SCHED_PROCESSING_READ_HISTOGRAM_VEC.local(),
+            command_keyreads: KV_COMMAND_KEYREAD_HISTOGRAM_VEC.local(),
+            kv_command_counter: KV_COMMAND_COUNTER_VEC.local(),
+            raw_command_counter: RAWKV_COMMAND_COUNTER_VEC.local(),
+            command_pri_counter: SCHED_COMMANDS_PRI_COUNTER_VEC.local(),
+            scan_details: KV_COMMAND_SCAN_DETAILS.local(),
+            read_flow_stats: HashMap::default(),
+        }
+    }
+
+    #[inline]
+    pub fn collect_command_count_and_duration(
```

Comment:
it is strange to return a Timer for the collect function.  You may use another get function.

Reply:
how about `start_command_duration_timer`?

Reply:
ok